lib/HeroBonus.cpp

```diff
@@ -1269,15 +1338,26 @@ void CBonusSystemNode::removedRedDescendant(CBonusSystemNode *descendant)
 		if(b->propagator)
 			descendant->unpropagateBonus(b);
 
-	FOREACH_RED_PARENT(parent)
-		parent->removedRedDescendant(descendant);
+	TNodes redParents;
+	getRedAncestors(redParents); //get all red parents recursively

```

Comment:
Why `newRedDescendant` have special case for TOWN but this function not?

Reply:
That case should be only applicable for battles (rewritten), so on. Battle object will be deleted.